src/storage/src/hummock/store/memtable.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::RangeBounds;
+
+use bytes::Bytes;
+
+pub trait Memtable: Send + Sync + 'static + Clone {
+    type Iter;
+
+    /// Inserts a key-value entry associated with a given `epoch` into memtable.
+    fn insert(&mut self, key: Bytes, val: Bytes, epoch: u64);
```

Comment:
> Given that the memtable will only be written by one actor, the written epoch should be increasing monotonically. Therefore, I will suggest we don't pass the `epoch` in `insert` and `delete`, and instead we introduce a new method named like `advance_epoch` to explicitly advance the current epoch.

if i understand correctly, is not easy to maintain. it may means two compose to advance `epoch` right ? (barrier_manager and memtable)

Reply:
Then we will need `advance_epoch` for both memtable and state store. I would rather avoid introducing avoid introducing one more interface for a specific assumption because 1) we need to call `advence_epoch` on barrier, which is very easy to forget; 2) upper layer can already guarantee monotonically increasing epoch.

However, it is okay for the implementation to add a sanity check to make sure the epoch never goes backward.

Reply:
Discussed offline. `advance_epoch` is useful because storage can know when an epoch is "sealed" (i.e. no future writes with this epoch). This can help the future queryable state store for batch query to know when it is ready to served in-memory state. Also, this information can also be useful to guide state store to do better batching and flush-ahead.

Therefore, we decided to add `advance_epoch` in the interface and remove `epoch` in `insert/delete`.